elastic/beats
# [Ingest Manager] Support for UPGRADE_ACTION

## What does this PR do?

Implements UPGRADE_ACTION defined in https://github.com/elastic/kibana/issues/73069

Once action is invoked from fleet agent will initiate an update procedure
this consists of few steps:
- **download**: fetches specified binaries from source specified by fleet
- **verify**: agent verifies package is created and signed by elastic
- **unpack**: agent unpack bits to correct location
- **switch** and **mark**: agent switches symlinks to point to correct version, updates paths and generates file describing upgrade
- **reexec**: agent reexecs itself so it is restarted in a new version

this PR does not add rollback capability

## Why is it important?

Sets the cornerstone for upgrade process initiated by fleet

## How to test
- build package
- copy them to some other location e.g `~/server/beats/elastic-agent`
- make a small change and commit (so commit inside a package changes and it is a valid upgrade)
- run `mage package` again
- unpack *.tar.gz
- run ES and KBN from branch in this PR https://github.com/elastic/kibana/pull/77412
- enroll & run agent
- run `python3 -m http.server` inside `~/server` so you can fetch update capable agent
- execute and replace AGENT_ID, this will trigger update
```
curl -X POST -H 'kbn-xsrf: true'  -u elastic:changeme http://localhost:5601/api/ingest_manager/fleet/agents/AGENT_ID/upgrade \
  -H 'Content-Type: application/json' \
 -d '{"source_uri":"http://localhost:8000","version":"8.0.0"}'
```
- check you have new agent inside data, it is running and action is ACKed

## Checklist
